WalletWasabi/WabiSabi/Client/CoinJoinClient.cs

```diff
@@ -235,12 +226,39 @@ public async Task<CoinJoinResult> StartRoundAsync(IEnumerable<SmartCoin> smartCo
 				aliceClientAndCircuit.AliceClient.Finish();
 				aliceClientAndCircuit.PersonCircuit.Dispose();
 			}
-			InCriticalCoinJoinState = false;
+			CoinJoinClientProgress.SafeInvoke(this, new LeavingCriticalPhase());
 		}
 	}
 
 	private async Task<ImmutableArray<(AliceClient AliceClient, PersonCircuit PersonCircuit)>> CreateRegisterAndConfirmCoinsAsync(IEnumerable<SmartCoin> smartCoins, RoundState roundState, CancellationToken cancellationToken)
 	{
+		bool alreadyEnteredToCritical = false;
+		object alreadyEnteredToCriticalLock = new();
+
+		void ReportProgressOnce()
+		{
+			if (alreadyEnteredToCritical)
+			{
+				return;
+			}
+
+			// Helper to not invoke event inside the lock.
+			bool reportProgress = false;
+			lock (alreadyEnteredToCriticalLock)
+			{
+				if (alreadyEnteredToCritical)
+				{
+					return;
+				}
```

Comment:
Is this check necessary?

Reply:
Yes it is the double lock pattern https://codeql.github.com/codeql-query-help/csharp/cs-unsafe-double-checked-lock/